In [178]:
import pandas as pd

In [179]:
train_data_file = '/Users/mackbookair/Desktop/Git/CapStone/CapstoneProject/TestData/allstate-purchase-prediction-challenge-datawargling/train.csv'

In [180]:
train_df = pd.read_csv(train_data_file)

In [181]:
train_df.head()

customer_ID  shopping_pt  record_type  day   time state  location  \
0     10000000            1            0    0  08:35    IN     10001   
1     10000000            2            0    0  08:38    IN     10001   
2     10000000            3            0    0  08:38    IN     10001   
3     10000000            4            0    0  08:39    IN     10001   
4     10000000            5            0    0  11:55    IN     10001   

   group_size  homeowner  car_age  ... C_previous  duration_previous  A  B  C  \
0           2          0        2  ...        1.0                2.0  1  0  2   
1           2          0        2  ...        1.0                2.0  1  0  2   
2           2          0        2  ...        1.0                2.0  1  0  2   
3           2          0        2  ...        1.0                2.0  1  0  2   
4           2          0        2  ...        1.0                2.0  1  0  2   

   D  E  F  G  cost  
0  2  1  2  2   633  
1  2  1  2  1   630  
2  2  1  2  1   630  
3  2  1  2  1   630  
4  2  1  2  1   630  

[5 rows x 25 columns]

In [182]:
train_df.columns

Index(['customer_ID', 'shopping_pt', 'record_type', 'day', 'time', 'state',
       'location', 'group_size', 'homeowner', 'car_age', 'car_value',
       'risk_factor', 'age_oldest', 'age_youngest', 'married_couple',
       'C_previous', 'duration_previous', 'A', 'B', 'C', 'D', 'E', 'F', 'G',
       'cost'],
      dtype='object')

In [183]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 665249 entries, 0 to 665248
Data columns (total 25 columns):
customer_ID          665249 non-null int64
shopping_pt          665249 non-null int64
record_type          665249 non-null int64
day                  665249 non-null int64
time                 665249 non-null object
state                665249 non-null object
location             665249 non-null int64
group_size           665249 non-null int64
homeowner            665249 non-null int64
car_age              665249 non-null int64
car_value            663718 non-null object
risk_factor          424831 non-null float64
age_oldest           665249 non-null int64
age_youngest         665249 non-null int64
married_couple       665249 non-null int64
C_previous           646538 non-null float64
duration_previous    646538 non-null float64
A                    665249 non-null int64
B                    665249 non-null int64
C                    665249 non-null int64
D                   

Observations:
1. 665249 records
2. 25 columns
3. time column is identified as object
3. Following columns have NA -
    car_value,
    risk_factor,
    C_previous,
    duration_previous
4. column name 'customer_ID' is inconsistent


In [248]:
# changing the customer_ID column name to customer_id
train_df.rename(columns={'customer_ID':'customer_id'},inplace = True)

In [249]:
train_df.columns

Index(['customer_id', 'shopping_pt', 'record_type', 'day', 'time', 'state',
       'location', 'group_size', 'homeowner', 'car_age', 'car_value',
       'risk_factor', 'age_oldest', 'age_youngest', 'married_couple',
       'C_previous', 'duration_previous', 'A', 'B', 'C', 'D', 'E', 'F', 'G',
       'cost'],
      dtype='object')

In [186]:
#dropping the na rows
train_df_clean = train_df.dropna()

In [187]:
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
customer_id          417447 non-null int64
shopping_pt          417447 non-null int64
record_type          417447 non-null int64
day                  417447 non-null int64
time                 417447 non-null object
state                417447 non-null object
location             417447 non-null int64
group_size           417447 non-null int64
homeowner            417447 non-null int64
car_age              417447 non-null int64
car_value            417447 non-null object
risk_factor          417447 non-null float64
age_oldest           417447 non-null int64
age_youngest         417447 non-null int64
married_couple       417447 non-null int64
C_previous           417447 non-null float64
duration_previous    417447 non-null float64
A                    417447 non-null int64
B                    417447 non-null int64
C                    417447 non-null int64
D                   

In [219]:
train_df_clean = train_df_clean.drop_duplicates()

In [220]:
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
customer_id          417447 non-null int64
shopping_pt          417447 non-null int64
record_type          417447 non-null int64
day                  417447 non-null int64
time                 417447 non-null object
state                417447 non-null category
location             417447 non-null int64
group_size           417447 non-null int64
homeowner            417447 non-null int64
car_age              417447 non-null int64
car_value            417447 non-null object
risk_factor          417447 non-null float64
age_oldest           417447 non-null int64
age_youngest         417447 non-null int64
married_couple       417447 non-null category
C_previous           417447 non-null float64
duration_previous    417447 non-null float64
A                    417447 non-null int64
B                    417447 non-null int64
C                    417447 non-null int64
D              

No duplicates were observed

In [250]:
train_df_clean.risk_factor.unique()

array([3., 4., 2., 1.])

In [253]:
train_df_clean['risk_factor'] = train_df_clean['risk_factor'].astype('int')

In [256]:
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
customer_id          417447 non-null int64
shopping_pt          417447 non-null int64
record_type          417447 non-null int64
day                  417447 non-null int64
time                 417447 non-null object
state                417447 non-null category
location             417447 non-null int64
group_size           417447 non-null int64
homeowner            417447 non-null int64
car_age              417447 non-null int64
car_value            417447 non-null object
risk_factor          417447 non-null int64
age_oldest           417447 non-null int64
age_youngest         417447 non-null int64
married_couple       417447 non-null category
C_previous           417447 non-null float64
duration_previous    417447 non-null float64
A                    417447 non-null int64
B                    417447 non-null int64
C                    417447 non-null int64
D                

In [192]:
train_df_clean.C_previous.unique()

array([1., 3., 2., 4.])

In [257]:
train_df_clean['C_previous'] = train_df_clean['C_previous'].astype('int')
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
customer_id          417447 non-null int64
shopping_pt          417447 non-null int64
record_type          417447 non-null int64
day                  417447 non-null int64
time                 417447 non-null object
state                417447 non-null category
location             417447 non-null int64
group_size           417447 non-null int64
homeowner            417447 non-null int64
car_age              417447 non-null int64
car_value            417447 non-null object
risk_factor          417447 non-null int64
age_oldest           417447 non-null int64
age_youngest         417447 non-null int64
married_couple       417447 non-null category
C_previous           417447 non-null int64
duration_previous    417447 non-null float64
A                    417447 non-null int64
B                    417447 non-null int64
C                    417447 non-null int64
D                  

In [193]:
train_df_clean.duration_previous.unique()

array([ 2., 13.,  3.,  0.,  1., 14.,  5., 15.,  4.,  6., 11.,  9., 10.,
        8.,  7., 12.])

In [258]:
train_df_clean['duration_previous']=train_df_clean['duration_previous'].astype('int')

In [259]:
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
customer_id          417447 non-null int64
shopping_pt          417447 non-null int64
record_type          417447 non-null int64
day                  417447 non-null int64
time                 417447 non-null object
state                417447 non-null category
location             417447 non-null int64
group_size           417447 non-null int64
homeowner            417447 non-null int64
car_age              417447 non-null int64
car_value            417447 non-null object
risk_factor          417447 non-null int64
age_oldest           417447 non-null int64
age_youngest         417447 non-null int64
married_couple       417447 non-null category
C_previous           417447 non-null int64
duration_previous    417447 non-null int64
A                    417447 non-null int64
B                    417447 non-null int64
C                    417447 non-null int64
D                    

In [260]:
train_df_clean.A.unique()

array([1, 0, 2])

In [261]:
len(train_df_clean.A.unique())

3

In [262]:
train_df_clean.B.unique()

array([0, 1])

In [263]:
train_df_clean.C.unique()

array([2, 3, 1, 4])

In [264]:
train_df_clean.D.unique()

array([2, 3, 1])

In [265]:
train_df_clean.E.unique()

array([1, 0])

In [266]:
train_df_clean.F.unique()

array([2, 0, 1, 3])

In [267]:
train_df_clean.G.unique()

array([2, 1, 3, 4])

In [268]:
train_df_clean.C_previous.unique()

array([1, 3, 2, 4])

In [269]:
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
customer_id          417447 non-null int64
shopping_pt          417447 non-null int64
record_type          417447 non-null int64
day                  417447 non-null int64
time                 417447 non-null object
state                417447 non-null category
location             417447 non-null int64
group_size           417447 non-null int64
homeowner            417447 non-null int64
car_age              417447 non-null int64
car_value            417447 non-null object
risk_factor          417447 non-null int64
age_oldest           417447 non-null int64
age_youngest         417447 non-null int64
married_couple       417447 non-null category
C_previous           417447 non-null int64
duration_previous    417447 non-null int64
A                    417447 non-null int64
B                    417447 non-null int64
C                    417447 non-null int64
D                    

In [270]:
train_df_clean['car_value']

0         g
1         g
2         g
3         g
4         g
         ..
665232    g
665233    g
665240    f
665241    f
665242    g
Name: car_value, Length: 417447, dtype: object

In [271]:
train_df_clean.car_value.unique()

array(['g', 'e', 'd', 'f', 'c', 'h', 'i', 'b', 'a'], dtype=object)

In [272]:
train_df_clean['car_value'].value_counts()

e    135026
f    113445
d     69283
g     63481
h     20106
c     11911
i      2867
b       712
a       616
Name: car_value, dtype: int64

In [274]:
train_df_clean['car_value'] = train_df_clean['car_value'].astype('category')
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
customer_id          417447 non-null int64
shopping_pt          417447 non-null int64
record_type          417447 non-null int64
day                  417447 non-null int64
time                 417447 non-null object
state                417447 non-null category
location             417447 non-null int64
group_size           417447 non-null int64
homeowner            417447 non-null int64
car_age              417447 non-null int64
car_value            417447 non-null category
risk_factor          417447 non-null int64
age_oldest           417447 non-null int64
age_youngest         417447 non-null int64
married_couple       417447 non-null category
C_previous           417447 non-null int64
duration_previous    417447 non-null int64
A                    417447 non-null int64
B                    417447 non-null int64
C                    417447 non-null int64
D                  

In [275]:
train_df_clean.state.unique()

[IN, NY, WV, MO, PA, ..., MT, NH, DC, WY, SD]
Length: 36
Categories (36, object): [IN, NY, WV, MO, ..., NH, DC, WY, SD]

In [276]:
train_df_clean['state'].value_counts()

FL    103762
NY     89334
MD     27557
PA     26194
IN     23675
OH     15656
WA     13101
CO     10976
AL      9883
OR      7793
TN      7193
CT      7139
NV      6248
ID      5748
UT      5603
MO      5516
KY      5228
OK      4728
WI      4265
GA      4028
DE      3972
MS      3411
NH      3277
AR      3154
RI      3095
NM      2999
KS      2442
ME      2275
WV      2080
IA      1697
NE      1604
MT      1511
DC      1078
WY       563
ND       456
SD       206
Name: state, dtype: int64

In [277]:
train_df_clean['day'].value_counts()

0    90026
1    86350
2    83883
3    76308
4    74723
5     5442
6      715
Name: day, dtype: int64

In [278]:
train_df_clean['location'].value_counts()

10213    974
11517    864
10030    781
12091    730
10793    693
        ... 
14702      1
16026      1
15220      1
15862      1
15199      1
Name: location, Length: 5888, dtype: int64

In [279]:
train_df_clean['group_size'].value_counts()

1    320197
2     90918
3      5893
4       439
Name: group_size, dtype: int64

In [280]:
train_df_clean['homeowner'].value_counts()

1    248041
0    169406
Name: homeowner, dtype: int64

In [281]:
train_df_clean['car_age'].value_counts()

1     50441
2     34040
7     28664
6     28180
8     27678
      ...  
85        4
59        3
51        2
62        1
53        1
Name: car_age, Length: 66, dtype: int64

In [282]:
train_df_clean['car_value'].value_counts()

e    135026
f    113445
d     69283
g     63481
h     20106
c     11911
i      2867
b       712
a       616
Name: car_value, dtype: int64

In [283]:
train_df_clean['risk_factor'].value_counts()

3    114562
4    108632
1     98472
2     95781
Name: risk_factor, dtype: int64

In [284]:
train_df_clean['age_oldest'].value_counts()

75    40029
25    11944
24    11232
23    10687
26    10645
22     9208
27     9069
28     8461
66     7678
29     7555
55     7501
65     7500
63     7198
30     7133
58     7057
56     7027
59     7016
31     6986
60     6845
62     6830
32     6788
61     6732
70     6717
64     6714
52     6679
49     6675
54     6631
21     6601
53     6597
33     6581
57     6543
51     6478
48     6374
43     6326
50     6171
46     6160
42     6133
67     6088
69     6087
34     5995
44     5901
35     5864
45     5843
71     5786
41     5619
47     5585
68     5503
37     5467
36     5412
40     5396
39     5171
72     5151
38     5083
74     4945
73     4888
20     4582
19     3426
18     1154
Name: age_oldest, dtype: int64

In [285]:
train_df_clean['age_youngest'].value_counts()

75    30795
25    13303
24    12458
23    12124
26    11768
22    10750
27    10053
28     9567
29     8373
21     8364
30     7856
31     7801
32     7412
65     7292
66     7122
55     7043
33     6856
58     6627
62     6558
63     6545
60     6536
56     6453
61     6384
34     6349
35     6327
20     6322
64     6302
59     6243
52     6236
57     6223
54     6155
53     6116
51     6110
49     6059
70     5997
48     5924
42     5806
69     5756
50     5745
46     5732
43     5726
44     5552
37     5527
45     5520
41     5406
36     5396
71     5380
40     5366
38     5320
19     5296
68     5224
39     5186
47     5104
67     5050
72     4463
73     4451
74     4422
18     3149
16     2676
17     1791
Name: age_youngest, dtype: int64

In [286]:
train_df_clean['married_couple'].value_counts()

0    322944
1     94503
Name: married_couple, dtype: int64

In [287]:
train_df_clean['duration_previous'].value_counts()

15    64736
2     44079
1     42179
3     41620
4     35712
5     32225
6     30164
7     26330
8     21591
9     19175
10    13765
0     12800
11     9480
12     8204
13     8149
14     7238
Name: duration_previous, dtype: int64

In [288]:
train_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 417447 entries, 0 to 665242
Data columns (total 25 columns):
customer_id          417447 non-null int64
shopping_pt          417447 non-null int64
record_type          417447 non-null int64
day                  417447 non-null int64
time                 417447 non-null object
state                417447 non-null category
location             417447 non-null int64
group_size           417447 non-null int64
homeowner            417447 non-null int64
car_age              417447 non-null int64
car_value            417447 non-null category
risk_factor          417447 non-null int64
age_oldest           417447 non-null int64
age_youngest         417447 non-null int64
married_couple       417447 non-null category
C_previous           417447 non-null int64
duration_previous    417447 non-null int64
A                    417447 non-null int64
B                    417447 non-null int64
C                    417447 non-null int64
D                  

In [289]:
train_df_clean['married_couple'] = train_df_clean['married_couple'].astype('category')

In [290]:
train_df_clean['state'] = train_df_clean['state'].astype('category')

In [291]:
print(train_df_clean.dtypes)

customer_id             int64
shopping_pt             int64
record_type             int64
day                     int64
time                   object
state                category
location                int64
group_size              int64
homeowner               int64
car_age                 int64
car_value            category
risk_factor             int64
age_oldest              int64
age_youngest            int64
married_couple       category
C_previous              int64
duration_previous       int64
A                       int64
B                       int64
C                       int64
D                       int64
E                       int64
F                       int64
G                       int64
cost                    int64
dtype: object


In [292]:
train_df_clean['married_couple']

0         1
1         1
2         1
3         1
4         1
         ..
665232    0
665233    0
665240    0
665241    0
665242    0
Name: married_couple, Length: 417447, dtype: category
Categories (2, int64): [0, 1]

In [293]:
train_df_clean['state']

0         IN
1         IN
2         IN
3         IN
4         IN
          ..
665232    FL
665233    FL
665240    FL
665241    FL
665242    FL
Name: state, Length: 417447, dtype: category
Categories (36, object): [AL, AR, CO, CT, ..., WA, WI, WV, WY]

In [294]:
len(train_df_clean[train_df_clean.record_type==1])

62255

In [300]:
len(train_df_clean[train_df_clean.record_type==1].customer_id.unique())

62255

In [295]:
len(train_df_clean.customer_id.unique())

62978

In [303]:
#l = train_df_clean[train_df_clean.record_type==1]
#l

only 62978 records have purchased coverage options out of 62978 unique customer_ids

In [304]:
train_df_p = train_df_clean[train_df_clean['customer_id'].isin(train_df_clean[train_df_clean.record_type==1].customer_id)]
train_df_p

customer_id  shopping_pt  record_type  day   time state  location  \
0          10000000            1            0    0  08:35    IN     10001   
1          10000000            2            0    0  08:38    IN     10001   
2          10000000            3            0    0  08:38    IN     10001   
3          10000000            4            0    0  08:39    IN     10001   
4          10000000            5            0    0  11:55    IN     10001   
...             ...          ...          ...  ...    ...   ...       ...   
665232     10152720            7            0    3  15:24    FL     14100   
665233     10152720            8            1    4  08:34    FL     14100   
665240     10152723            1            0    2  15:13    FL     10711   
665241     10152723            2            0    2  15:14    FL     10711   
665242     10152723            3            1    1  10:30    FL     10711   

        group_size  homeowner  car_age  ... C_previous  duration_previous  A  \
0                2          0        2  ...          1                  2  1   
1                2          0        2  ...          1                  2  1   
2                2          0        2  ...          1                  2  1   
3                2          0        2  ...          1                  2  1   
4                2          0        2  ...          1                  2  1   
...            ...        ...      ...  ...        ...                ... ..   
665232           1          0       13  ...          1                  8  1   
665233           1          0       13  ...          1                  8  1   
665240           1          1        0  ...          3                  2  0   
665241           1          1        0  ...          3                  2  1   
665242           1          1        0  ...          3                  7  1   

        B  C  D  E  F  G  cost  
0       0  2  2  1  2  2   633  
1       0  2  2  1  2  1   630  
2       0  2  2  1  2  1   630  
3       0  2  2  1  2  1   630  
4       0  2  2  1  2  1   630  
...    .. .. .. .. .. ..   ...  
665232  0  1  2  0  2  3   681  
665233  0  1  2  0  2  3   681  
665240  0  2  1  0  0  4   656  
665241  0  3  3  0  2  3   687  
665242  0  3  3  1  2  3   651  

[414812 rows x 25 columns]

In [305]:
len(train_df_p.customer_id.unique())

62255

In [306]:
train_df_p.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 414812 entries, 0 to 665242
Data columns (total 25 columns):
customer_id          414812 non-null int64
shopping_pt          414812 non-null int64
record_type          414812 non-null int64
day                  414812 non-null int64
time                 414812 non-null object
state                414812 non-null category
location             414812 non-null int64
group_size           414812 non-null int64
homeowner            414812 non-null int64
car_age              414812 non-null int64
car_value            414812 non-null category
risk_factor          414812 non-null int64
age_oldest           414812 non-null int64
age_youngest         414812 non-null int64
married_couple       414812 non-null category
C_previous           414812 non-null int64
duration_previous    414812 non-null int64
A                    414812 non-null int64
B                    414812 non-null int64
C                    414812 non-null int64
D                  